### PR TITLE
Initial commit, issue-9908

### PR DESCRIPTION
**Pull request by samlitowitz**
Addresses #9908 

Composer will use headers provided as part of repository package definintions.

### CODE REVIEW

File: src/Composer/Util/Http/CurlDownloader.php [hunk 1]
```diff
@@ -29,7 +30,7 @@
  */
 class CurlDownloader
 {
-    private $multiHandle;
+    public $multiHandle;
```
Comment (stof) [inline src/Composer/Util/Http/CurlDownloader.php:33]: why making this public ?

File: src/Composer/Util/Http/HeaderHelper.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+        foreach ($repos as $repo) {
+            if ($repo['type'] !== 'package') {
+                continue;
+            }
+            if (!array_key_exists('package', $repo)) {
+                continue;
+            }
+            if (
+                array_key_exists('dist', $repo['package'])
```
Comment (stof) [inline src/Composer/Util/Http/HeaderHelper.php:71]: a `package` repository can provide multiple packages

File: src/Composer/Util/Http/CurlDownloader.php [hunk 3]
```diff
@@ -184,6 +189,7 @@ private function initDownload($resolve, $reject, $origin, $url, $options, $copyT
 
         $options['http']['header'] = $this->authHelper->addAuthenticationHeader($options['http']['header'], $origin, $url);
         $options = StreamContextFactory::initOptions($url, $options, true);
+        $options['http']['header'] = $this->headerHelper->addHeader($options['http']['header'], $origin, $url);
```
Comment (stof) [inline src/Composer/Util/Http/CurlDownloader.php:192]: This implementation looks weird to me:

- it works only for the CurlDownloader
- it goes reading the configuration from repositories of type `package` (why only those ?) to alter headers even for URLs which are not related to that repository (well, they have to match).

to me, it looks like this has been implemented in the wrong layer of composer.

File: src/Composer/Util/Http/HeaderHelper.php (hunk 2 above)
Comment (samlitowitz) [reply src/Composer/Util/Http/HeaderHelper.php:71]: A complication for sure, probably address this after the discussion where to implement this if it remains relevant.

File: src/Composer/Util/Http/CurlDownloader.php (hunk 1 above)
Comment (samlitowitz) [reply src/Composer/Util/Http/CurlDownloader.php:33]: Unintentional.

File: src/Composer/Util/Http/CurlDownloader.php (hunk 3 above)
Comment (samlitowitz) [reply src/Composer/Util/Http/CurlDownloader.php:192]: I am unfamiliar with the composer architecture and welcome suggestions. Only applying it to `package` types does appear to be unnecessarily limiting.

**Pull request closed without merging**